src/bitmap/bytes.rs

```diff
@@ -0,0 +1,255 @@
+use crate::bitmap::{bitmask_for_key, index_for_key};
+use crate::Bitmap;
+use bytes::{BufMut, Bytes, BytesMut};
+use std::convert::TryInto;
+
+/// A plain, heap-allocated, `O(1)` indexed bitmap using `bytes::BytesMut` for storage.
+///
+/// This bitmap requires `O(n)` space and can be read and wrote to in `O(1)` time.
+///
+/// This type is fast for both reads and writes, trading internal complexity for speed.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct BytesBitmap {
+    max_key: usize,
+    bitmap: BytesMut,
+}
+
+impl BytesBitmap {
+    pub(crate) fn into_parts(self) -> (Vec<usize>, usize) {
+        let mut vec = Vec::with_capacity(self.bitmap.len() / size_of::<usize>());
+        let chunks = self.bitmap.chunks_exact(size_of::<usize>());
+        for chunk in chunks {
+            vec.push(usize::from_ne_bytes(chunk.try_into().unwrap()));
+        }
+        (vec, self.max_key)
+    }
+
+    pub(crate) fn shrink(&mut self) {
+        self.bitmap.resize(self.bitmap.len(), 0);
+    }
+
+    pub fn bitmap(self) -> Bytes {
+        self.bitmap.freeze()
+    }
+
+    pub fn max_key(&self) -> usize {
+        self.max_key
+    }
+
+    pub fn from_bytes(bitmap: Bytes, max_key: usize) -> Self {
+        Self {
+            max_key,
+            bitmap: BytesMut::from(bitmap),
+        }
+    }
+}
+
+impl Bitmap for BytesBitmap {
+    fn new_with_capacity(max_key: usize) -> Self {
+        let size = (index_for_key(max_key) + 1) * size_of::<usize>();
+        let mut bytes = BytesMut::zeroed(size);
+
+        Self {
+            bitmap: bytes,
+            max_key,
+        }
+    }
+
+    fn set(&mut self, key: usize, value: bool) {
+        let offset = index_for_key(key);
+        let byte_offset = offset * size_of::<usize>();
+
+        let slice = &mut self.bitmap[byte_offset..byte_offset + size_of::<usize>()];
+        let mut num = usize::from_ne_bytes(slice.try_into().unwrap());
+
+        if value {
+            num |= bitmask_for_key(key);
+        } else {
+            num &= !bitmask_for_key(key);
+        }
```

Comment:
i think this is one of those "mental complexity vs raw performance" challenges where i opted for "can someone else understand this later?". i think the biggest penalty here is the inevitable branch.

i did get curious tho, so i tested it with:

```
// old
if value {
    num |= bitmask_for_key(key);
} else {
    num &= !bitmask_for_key(key);
}

// new oneline
let bitmask = 1 << (key % (u64::BITS as usize));
num = (num & !bitmask) | (bitmask & (value as usize));
```

and this is what it looks like:
```
bytes_bitmap_insert_true
                        time:   [804.32 ps 806.40 ps 808.66 ps]
                        change: [-3.6809% -2.9364% -2.1772%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

bytes_bitmap_insert_false
                        time:   [831.44 ps 842.49 ps 855.27 ps]
                        change: [+1.9304% +2.9006% +4.0927%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  8 (8.00%) high mild
  2 (2.00%) high severe

bytes_bitmap_lookup_hit time:   [816.60 ps 822.18 ps 828.08 ps]
                        change: [+0.5063% +0.9796% +1.4763%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

bytes_bitmap_lookup_miss, different block
                        time:   [804.50 ps 806.31 ps 808.19 ps]
                        change: [-0.9178% -0.4502% +0.0686%] (p = 0.06 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

bytes_bitmap_lookup miss, same block
                        time:   [809.17 ps 814.55 ps 821.69 ps]
                        change: [-1.0797% -0.4035% +0.3524%] (p = 0.29 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe
```

i think this is a case where trying to be smarter than the compiler is 6 of one, half dozen of a another because of the inevitable branch. i'll leave the oneline there, but realistically i don't think it matters much.